airflow/contrib/operators/gcp_compute_operator.py

```diff
@@ -0,0 +1,186 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow import AirflowException
+from airflow.contrib.hooks.gcp_compute_hook import GceHook
+from airflow.contrib.utils.field_validator import BodyFieldValidator
+from airflow.models import BaseOperator
+from airflow.utils.decorators import apply_defaults
+
+
+class GceInstanceStartOperator(BaseOperator):
+    """
+    Start an instance in Google Compute Engine.
+
+    :param project_id: Google Cloud Platform project where the Compute Engine
+        instance exists.
+    :type project_id: str
+    :param zone: Google Cloud Platform zone where the instance exists.
+    :type zone: str
+    :param resource_id: Name of the Compute Engine instance resource.
+    :type resource_id: str
+    :param gcp_conn_id: The connection ID used to connect to Google Cloud Platform.
+    :type gcp_conn_id: str
+    :param api_version: API version used (e.g. v1).
+    :type api_version: str
+    """
+    @apply_defaults
+    def __init__(self,
+                 project_id,
+                 zone,
+                 resource_id,
+                 gcp_conn_id='google_cloud_default',
+                 api_version='v1',
+                 *args, **kwargs):
+        self.project_id = project_id
+        self.zone = zone
+        self.full_location = 'projects/{}/zones/{}'.format(self.project_id,
+                                                           self.zone)
+        self.resource_id = resource_id
+        self.gcp_conn_id = gcp_conn_id
+        self.api_version = api_version
+        self._validate_inputs()
+        self._hook = GceHook(gcp_conn_id=self.gcp_conn_id, api_version=self.api_version)
+        super(GceInstanceStartOperator, self).__init__(*args, **kwargs)
+
+    def _validate_inputs(self):
+        if not self.project_id:
+            raise AirflowException("The required parameter 'project_id' is missing")
+        if not self.zone:
+            raise AirflowException("The required parameter 'zone' is missing")
+        if not self.resource_id:
+            raise AirflowException("The required parameter 'resource_id' is missing")
+
+    def execute(self, context):
+        return self._hook.start_instance(self.project_id, self.zone, self.resource_id)
```

Comment:
Have you confirmed that this is idempotent? E.g. if the operator crashes, but the instance is created, subsequent runs succeed

Reply:
In 2d059a0e752b872c6db10e446b43364895bef877 I've enhanced the example DAG by duplicating each task (`start` >> `start` >> `stop` >> `stop` >> `setMachineType` >> `setMachineType`) to test idempotence.
I ran this as an integration test against GCP and everything worked fine, so all 3 operations are idempotent.
Let me know if you think it is enough of a guarantee.

Reply:
SGTM